crates/ink/src/codegen/dispatch/execution.rs

```diff
@@ -146,37 +133,38 @@ pub trait ConstructorReturnType<C>: private::Sealed {
     /// For infallible constructor returns this always yields `()`
     /// and is basically ignored since this does not get called
     /// if the constructor did not fail.
-    fn return_value(&self) -> &Self::ReturnValue;
+    fn error_value(&self) -> &Self::Error;
 }
 
 impl<C> ConstructorReturnType<C> for private::Seal<C> {
-    type Error = Infallible;
-    type ReturnValue = ();
+    type Error = ();
 
     #[inline]
     fn as_result(&self) -> Result<&C, &Self::Error> {
         Ok(&self.0)
     }
 
     #[inline]
-    fn return_value(&self) -> &Self::ReturnValue {
+    fn error_value(&self) -> &Self::Error {
         &()
     }
 }
 
 impl<C, E> ConstructorReturnType<C> for private::Seal<Result<C, E>> {
     const IS_RESULT: bool = true;
     type Error = E;
-    type ReturnValue = Result<C, E>;
 
     #[inline]
     fn as_result(&self) -> Result<&C, &Self::Error> {
         self.0.as_ref()
     }
 
     #[inline]
-    fn return_value(&self) -> &Self::ReturnValue {
-        &self.0
+    fn error_value(&self) -> &Self::Error {
+        self.0
+            .as_ref()
+            .err()
+            .expect("value returned only when error occurs. qed")
```

Comment:
What exactly are we using this for? Could its uses be replaced by `as_result` instead? I don't like that it panics. When constructors were still infallible this API made sense but now I feel like it doesn't add too much value.

Reply:
We only return a *value* back to the caller when the constructor fails. In case of constructors, the return value is encoded error, that's why we explicitly unwrap the error. Hence, it makes more sense to rename the function to `error_value()` because it errors the values back to the caller.

This was a [suggestion from Andrew](https://github.com/paritytech/ink/pull/1446#discussion_r1004316076)

Reply:
See https://github.com/paritytech/ink/pull/1446#discussion_r1006621274.

Reply:
Yes it would be better just to match against the error value